### PR TITLE
charles/alignment-viz-2

### PR DESCRIPTION
**Pull request by cdebourcy**
# Description

Annotates fasta with full alignment info from m8 and makes that information available for display.
TO DO: update run_generate_unidentified_fasta and check if any other functions need to be mended

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix a feature that knowingly causes existing functionality to not work as expected)

# Test and Reproduce
1. process  or rerun a sample, determine its pipeline_output_id
2. pick one of the taxids that were hit, determine its tax_level, and go to http://localhost:3000/pipeline_outputs/:id/alignment_info/:tax_level/:taxid

## Impacted Areas in Application
List general components of the application that this PR will affect:
- pipeline
- pipeline_outputs/:id/alignment_info/:tax_level/:taxid page

# Checklist:

- [x] I have done relevant tests that prove my fix is effective or that my feature works
- [x] I have spent time testing out edge cases for my feature
- [x] I have updated the test script or pull request template if necessary
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/helpers/pipeline_outputs_helper.rb [hunk 1]
```diff
@@ -24,4 +24,47 @@ def get_s3_file(s3_path)
...
+    end
+    alignment_info
+  end
+
+  def get_sequence_length_from_accession(accession_id)
+    sequence_length = `
+      QUERY=#{accession_id}
+      BASE=https://eutils.ncbi.nlm.nih.gov/entrez/eutils
+      SEARCH_URL=${BASE}/esearch.fcgi?db=nuccore\&term=${QUERY}\&usehistory=y
+      OUTPUT=$(curl $SEARCH_URL)
```
Comment (cdebourcy) [inline app/helpers/pipeline_outputs_helper.rb]: this line is slow

File: app/helpers/pipeline_outputs_helper.rb [hunk 2]
```diff
@@ -24,4 +24,47 @@ def get_s3_file(s3_path)
...
+      WEB=$(echo $OUTPUT | sed -e 's/.*<WebEnv>\(.*\)<\/WebEnv>.*/\1/')
+      KEY=$(echo $OUTPUT | sed -e 's/.*<QueryKey>\(.*\)<\/QueryKey>.*/\1/')
+      FETCH_URL=${BASE}/efetch.fcgi?db=nuccore\&query_key=${KEY}\&WebEnv=${WEB}\&rettype=fasta\&retmode=xml
+      RESULT=$(curl $FETCH_URL)
```
Comment (cdebourcy) [inline app/helpers/pipeline_outputs_helper.rb]: this line is slow

File: app/controllers/pipeline_outputs_controller.rb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 class PipelineOutputsController < ApplicationController
   include ReportHelper
   include PipelineOutputsHelper
-  before_action :set_pipeline_output, only: [:show, :show_taxid_fasta, :send_nonhost_fasta, :send_unidentified_fasta]
+  before_action :set_pipeline_output, only: [:show, :show_taxid_fasta, :send_nonhost_fasta, :send_unidentified_fasta, :show_alignment_info]
```
Comment (yunfangjuan) [inline app/controllers/pipeline_outputs_controller.rb]: oops. pipeline_outputs_controller is gone.

File: app/controllers/pipeline_outputs_controller.rb [hunk 4]
```diff
@@ -41,6 +41,24 @@ def send_unidentified_fasta
     send_data @unidentified_fasta, filename: @pipeline_output.sample.name + '_unidentified.fasta'
   end
 
+  def show_alignment_info
+    hit_type = 'NT' # implementing NT first; to do: NR
```
Comment (yunfangjuan) [inline app/controllers/pipeline_outputs_controller.rb]: this could be moved to samples_controller easily.

File: aws_batch/non_host_alignment.py [hunk 5]
```diff
@@ -88,17 +88,28 @@
 AWS_BATCH_JOB_ID = None
 
 # convenience functions
-def generate_taxid_annotated_fasta_from_m8(input_fasta_file, m8_file, output_fasta_file, annotation_prefix):
+def fuzzy_get(dictionary, raw_key, default_value):
+    value_list = [dictionary[key] for key in dictionary if raw_key in key]
```
Comment (yunfangjuan) [inline aws_batch/non_host_alignment.py]: wouldn't this be kinda slow?

File: app/helpers/pipeline_outputs_helper.rb [hunk 6]
```diff
@@ -24,4 +24,49 @@ def get_s3_file(s3_path)
...
+    end
+    # Add reference length information (this uses the NCBI eutils API)
+    alignment_info.each do |accession_id, _info|
+      alignment_info[accession_id][:reference_length] = get_sequence_length_from_accession(accession_id)
+    end
+    alignment_info
+  end
+
+  def get_sequence_length_from_accession(accession_id)
+    sequence_length = `
```
Comment (yunfangjuan) [inline app/helpers/pipeline_outputs_helper.rb]: you are not getting the actual sequence just he sequence_length?

Comment (cdebourcy) [reply app/helpers/pipeline_outputs_helper.rb]: yeah, just need sequence length, but for that the code actually needs to get the full record including sequence unless we figure out a better way :/

File: aws_batch/non_host_alignment.py (hunk 5 above)
Comment (cdebourcy) [reply aws_batch/non_host_alignment.py]: yeah. It's only used for re-processing old samples, not new samples. I think the trouble was that old samples had annotations that made it impossible to know where the actual raw read-id started. Need to look into it again. Or suggestions welcome...

File: app/controllers/pipeline_outputs_controller.rb (hunk 3 above)
Comment (cdebourcy) [reply app/controllers/pipeline_outputs_controller.rb]: haha

**Pull request closed without merging**